SonarQube.Client/RoslynExporterAdapter/RoslynRuleSetGenerator.cs

```diff
@@ -0,0 +1,157 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using SonarQube.Client.Models;
+
+// Copied from the SonarScanner for MSBuild.
+// See https://github.com/SonarSource/sonar-scanner-msbuild/blob/9ccfdb648a0411014b29c7aee8e347aeab87ea71/src/SonarScanner.MSBuild.PreProcessor/Roslyn/RoslynRuleSetGenerator.cs#L28
+
+namespace SonarQube.Client.RoslynExporterAdapter
+{
+    public class RoslynRuleSetGenerator
+    {
+        private const string SONARANALYZER_PARTIAL_REPO_KEY = "sonaranalyzer-{0}";
+        private const string ROSLYN_REPOSITORY_PREFIX = "roslyn.";
+
+        private readonly IDictionary<string, string> sonarProperties;
+        private readonly string inactiveRuleActionText = GetActionText(RuleAction.None);
+
+        private string activeRuleActionText = GetActionText(RuleAction.Warning);
+
+        public RoslynRuleSetGenerator(IDictionary<string, string> sonarProperties)
+        {
+            this.sonarProperties = sonarProperties ?? throw new ArgumentNullException(nameof(sonarProperties));
+        }
+
+        /// <summary>
+        /// Generates a RuleSet that is serializable (XML).
+        /// The ruleset can be empty if there are no active rules belonging to the repo keys "vbnet", "csharpsquid" or "roslyn.*".
+        /// </summary>
+        /// <exception cref="InvalidOperationException">if required properties that should be associated with the repo key are missing.</exception>
+        public RuleSet Generate(string language, IEnumerable<SonarQubeRule> activeRules, IEnumerable<SonarQubeRule> inactiveRules)
+        {
+            if (activeRules == null)
+            {
+                throw new ArgumentNullException(nameof(activeRules));
+            }
+            if (inactiveRules == null)
+            {
+                throw new ArgumentNullException(nameof(inactiveRules));
+            }
+            if (language == null)
+            {
+                throw new ArgumentNullException(nameof(language));
+            }
+
+            var rulesElements = activeRules.Concat(inactiveRules)
```

Comment:
The ruleset generation is reasonably complex:
* rules can come from SonarC#, SonarVB, and from custom third-party analysis rules that have been wrapped by the SonarQube Roslyn SDK.
* the rules for each analyzer are grouped into a separate section in the ruleset e.g. 
```
<?xml version="1.0" encoding="utf-8"?>
<RuleSet xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Name="Rules for SonarQube" Description="This rule set was automatically generated from SonarQube" ToolsVersion="14.0">
  <Rules AnalyzerId="SonarAnalyzer.Security" RuleNamespace="SonarAnalyzer.Security">
    <Rule Id="S2631" Action="Warning" />
...
  </Rules>
  <Rules AnalyzerId="SonarAnalyzer.CSharp" RuleNamespace="SonarAnalyzer.CSharp">
    <Rule Id="S2225" Action="Warning" />
...
    <Rule Id="S1449" Action="None" />
  </Rules>
  <Rules AnalyzerId="Wintellect.Analyzers" RuleNamespace="Wintellect.Analyzers">
    <Rule Id="Wintellect007" Action="None" />
...
  </Rules>
</RuleSet>
```
* the generator works out which rules to include in the ruleset partly by looking for additional metadata it expects to be available via Sonar properties publishes by each analyzer/plugin. It also uses these properties to work out what names to use for the `RuleNamespace` etc.